datalad/support/gitrepo.py

```diff
@@ -3705,19 +3742,40 @@ def _get_cache_key(label, paths, ref, untracked=None):
 
         # loop over all subdatasets and look for additional modifications
         for f, st in status.items():
-            f = str(f)
-            if 'state' in st or not st['type'] == 'dataset':
-                # no business here
+            if st.get('state', None) or not st['type'] == 'dataset':
+                # no business here, we already know a state
                 continue
             if not GitRepo.is_valid_repo(f):
                 # submodule is not present, no chance for a conflict
                 st['state'] = 'clean'
                 continue
             # we have to recurse into the dataset and get its status
-            subrepo = GitRepo(f)
-            subrepo_commit = subrepo.get_hexsha()
+            subrepo = repo_from_path(f)
+            # figure out what we should compare against (in the face
+            # of a potential adjusted branch situation)
+            subm_cbranch = subrepo.get_corresponding_branch()
+            if subm_cbranch:
+                # this is actually an AnnexRepo in adjusted mode.
+                # now we have two choices:
+                # 1) run a diff of the adjusted branch against the
+                #    corresponding branch, and proceed with the
+                #    corresponding if the diff is clean, or label the
+                #    dataset 'modified' if not.
+                # 2) run an unconditional localsync() and therefor
+                #    enforce a clean diff between adjusted and
+                #    corresponding branch.
+                # Going with (2) due to implementation simplicity, but
+                # without having tested performance differences on complex
+                # datasets
+                lgr.debug(
+                    'Sync corresponding branch %s at %s prior repository '
+                    'state evaluation', subm_cbranch, f)
+                subrepo.localsync(managed_only=True)
```

Comment:
Can we be cheaper than this, especially if we just had sync'ed before? Should we make an attempt to count the commits on the adjusted branch, and if there is only one not go for a sync. This would only miss cases, where someone amends commits on the adjusted branch, but that would be problematic behavior anyways, so maybe that is good enough. OTOH, I dunno how clever git-annex is in figuring out that nothing needs to be done....

Reply:
I'd vote for sticking with the simpler `sync` call for now.

> OTOH, I dunno how clever git-annex is in figuring out that nothing needs to be done....

Quickly testing suggests that annex will _not_ sync back the changes if you amend the initial adjusted branch commit and have no other commits.